[MIG] l10n_ch_payment_slip: Migration to 11.0

yvaucher opened the pull request:
Update headers
Port code to python3
Update README
Fix tests with correct report render calls
Adapt code to report module split

TODO: 
- [x] fixing tests
- [x] v11 + tests -> moved in another module
- [x] put bvr parameters on res.confing.settings
- [ ] edit README.rst to explain the difference with Odoo official addon.

yvaucher committed:
[MIG] l10n_ch_payment_slip: Migration to 11.0

Update headers
Port code to python3
Update README
Fix tests with correct report render calls
Adapt code to report module split

grindtildeath reviewed (commented):
https://github.com/OCA/l10n-switzerland/pull/373/files#diff-788bc5e9535f52e4fdb14b3cf5d6a2e2R15

odoo.report doesn't exist in v11, so it crashes if submodule is in addons_path

```
2017-11-20 16:18:15,761 1 ERROR ? werkzeug: Error on request:
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/werkzeug/serving.py", line 193, in run_wsgi
    execute(self.server.app)
  File "/usr/local/lib/python3.5/dist-packages/werkzeug/serving.py", line 181, in execute
    application_iter = app(environ, start_response)
  File "/opt/odoo/src/odoo/service/server.py", line 251, in app
    return self.app(e, s)
  File "/opt/odoo/src/odoo/service/wsgi_server.py", line 166, in application
    return application_unproxied(environ, start_response)
  File "/opt/odoo/src/odoo/service/wsgi_server.py", line 154, in application_unproxied
    result = handler(environ, start_response)
  File "/opt/odoo/src/odoo/http.py", line 1297, in __call__
    self.load_addons()
  File "/opt/odoo/src/odoo/http.py", line 1319, in load_addons
    m = __import__('odoo.addons.' + module)
  File "/opt/odoo/src/odoo/modules/module.py", line 82, in load_module
    exec(open(modfile, 'rb').read(), new_mod.__dict__)
  File "<string>", line 1, in <module>
    
  File "/opt/odoo/external-src/l10n-switzerland/l10n_ch_payment_slip/models/__init__.py", line 2, in <module>
    from . import payment_slip
  File "/opt/odoo/external-src/l10n-switzerland/l10n_ch_payment_slip/models/payment_slip.py", line 15, in <module>
    from odoo.report import report_sxw
ImportError: No module named 'odoo.report'
```

yvaucher committed:
Rename BVR and BVR/ESR to ISR in english and make it the base term

grindtildeath committed:
Improve tests and assign bvr_address_format without sudo

grindtildeath committed:
Improve tests and assign bvr_address_format without sudo

simahawk commented on l10n_ch_payment_slip/__manifest__.py:
```diff
@@ -1,32 +1,28 @@
-# -*- coding: utf-8 -*-
-# © 2012-2016 Camptocamp SA
+# Copyright 2012-2016 Camptocamp SA
```
2012-2018, same for other py files

yvaucher replied:
When significant changes were made in py files I updated it already (was in 2017 BTW)

simahawk commented on l10n_ch_payment_slip/models/company.py:
```diff
@@ -1,62 +1,71 @@
 
 class ResCompany(models.Model):
-    """override company in order to add bvr vertical and
+    """override company in order to add ISR vertical and
```
```"""Add ISR vertical and Horizontal print delta."""```

simahawk commented on l10n_ch_payment_slip/models/invoice.py:
```diff
@@ -24,8 +23,8 @@ class AccountMoveLine(models.Model):
 class AccountInvoice(models.Model):
-    """Inherit account.invoice in order to add bvr
-    printing functionnalites. BVR is a Swiss payment vector"""
+    """Inherit account.invoice in order to add ISR
```
"""Add ISR (Swiss payment vector) printing features."""

simahawk commented on l10n_ch_payment_slip/report/ir_action.py:
```diff
@@ -1,46 +1,107 @@
+    @api.multi
+    def _generate_one_slip_per_page_from_invoice_pdf(self, report_name=None):
+        """Generate payment slip PDF(s) from report model.
+        If there is many pdf they are merged in memory or on
```
blank line after 1st one.

simahawk replied:
"If there are many PDF..."

simahawk reviewed (changes requested):
LGTM, just nitpicking non-blocking comments ;)

grindtildeath commented on l10n_ch_payment_slip/views/company.xml:
```diff
@@ -2,25 +2,25 @@
-            <field name="name">res.company.form.inherit.bvr</field>
+            <field name="name">res.company.form.inherit.isr</field>
             <field name="model">res.company</field>
             <field name="inherit_id" ref="base.view_company_form"/>
```
IMO would be nice to move these settings from the company from view to res.config.settings

Edit : Didn't see you todo ;)

grindtildeath commented on l10n_ch_payment_slip/__manifest__.py:
```diff
@@ -1,32 +1,28 @@
-     'report',
+     'account_invoicing',
      'l10n_ch_base_bank',
      'base_transaction_id',  # OCA/bank-statement-reconcile
```
OCA/account-reconcile

grindtildeath commented on l10n_ch_payment_slip/tests/test_payment_slip.py:
```diff
@@ -1,15 +1,19 @@
+            'l10n_ch_payment_slip.one_slip_per_page_from_invoice',
+        )
+
     def make_bank(self):
```
Would be great to have these records as demo data as it's impossible to test functionally without it.

grindtildeath reviewed (changes requested):
Maybe we need this commit ?
https://github.com/OCA/l10n-switzerland/pull/392/commits/266e82dd019d33f2cf1a948f347bdafe6b8fb614

Also I think we should add a dependency on contacts because that's where the res.bank and res.partner.bank menus are defined, and we need these to configure the module properly.

Moreover I get an error when I try to print ISR : 
```


Odoo Server Error

Traceback (most recent call last):
  File "/opt/odoo/src/addons/web/controllers/main.py", line 77, in wrap
    return f(*args, **kwargs)
  File "/opt/odoo/src/addons/web/controllers/main.py", line 1527, in index
    action["report_name"], report_ids, report_data, context])
  File "/opt/odoo/src/odoo/http.py", line 115, in dispatch_rpc
    result = dispatch(method, params)
UnboundLocalError: local variable 'dispatch' referenced before assignment
```

tschanzt commented on l10n_ch_payment_slip/models/company.py:
```diff
@@ -1,62 +1,71 @@
+    isr_delta_horz = fields.Float(
+        'ISR Horz. Delta (inch)',
+        oldname='bvr_delta_horz',
+        help='horiz. delta in inch 1.2 will print the ISR 1.2 inch lefter,'
```
lefter sounds weird. to the left?

tschanzt reviewed (commented):
👍 if you get travis to run.

simahawk committed:
l10n_ch_payment_slip: fix `_get_address_lines`

`payment_slip._get_address_lines` computes the address of the partner.

This is done in an onchange context whereas the changes where not isolated
and they were propagated to the real partner
via all the onchanges related to it.

We spotted the issue on the mail compose wizard
whereas the chain of onchanges was wiping all the default values.

With this change we:

* isolate the partner w/ a fake record
* additionally we cache the result per each partner
  since the function is called several times

grindtildeath committed:
Allow payment slip to be saved in and read from attachment

grindtildeath committed:
Delete payment slip attachment when setting invoice back to draft

grindtildeath committed:
Add test case on attachment management

yvaucher commented:
I'll close this PR as odoo is now managing the payment slip officially again.

If the official module lacks features we can still create other modules. On the other hand it might not be the right time to make lot of changes as the QR invoices will replace this.

TDu committed:
Fix some docstrings, typos and english language

TDu committed:
Fix test for python3

yvaucher commented:
Reopened as official module is currently not providing all required features. (alignment on printout, reconciliation based on BVR reference)

TDu committed:
Fix non qweb report not supported

TDu committed:
Remove v11 import

It has been moved to a own module l10n_ch_import_isr_v11

TDu committed:
Improve ISR file name

grindtildeath commented:
@TDu Can you please integrate the content of this PR as well ? https://github.com/OCA/l10n-switzerland/pull/411

TDu commented:
@grindtildeath It has been merged in 10.0, are you asking to add it to 11.0 ?

TDu committed:
Add python dependency

TDu committed:
Simplify js code

TDu commented:
@grindtildeath Will look at it when I get a green Travis.

TDu committed:
Fix batch print ISR

TDu committed:
Move parameters to setting view

TDu commented:
@grindtildeath Unfortunately it is not straightforward to incorporate your PR in 11.0.  Some methods have moved or been dropped, could you have a look at it ?

grindtildeath commented:
@TDu I cherry-picked and tried to fix but there's a few things not working yet, can you please check it as you already did most of the migration ? Thx
https://github.com/yvaucher/l10n-switzerland/pull/2

grindtildeath committed:
Fix v11 for attachment feature

Commit:
[11.0][IMP] Remove useless button on invoice

Remove the button "Print ISR" on invoices to avoid confusion with Odoo base
implementation and rename the one from this module for user consistency

yvaucher commented:
@grindtildeath thanks I added your changes in here.

simahawk commented:
reminder: squash before merging ;)

oleksandrpaziuk commented on l10n_ch_payment_slip/report/ir_action.py:
```diff
@@ -1,46 +1,201 @@
+                pdfdocuments.append(pdfreport_path)
+                continue
+
+            with closing(os.fdopen(pdfreport_fd, 'w')) as pdfreport:
```
@yvaucher can you replace `'w'` to `'wb'` or should I create PR in your repo?

yvaucher commented on l10n_ch_payment_slip/report/ir_action.py:
```diff
@@ -1,46 +1,201 @@
+
+            reports = self.browse(docids)
+            return reports._generate_one_slip_per_page_from_invoice_pdf(
+                report_name=report_name, save_in_attachment=save_in_attachment
```
report_name is undefined

yvaucher commented on l10n_ch_payment_slip/report/ir_action.py:
```diff
@@ -1,46 +1,201 @@
+            save_in_attachment = self._check_attachment_use(
+                docids, report)
+
+            reports = self.browse(docids)
```
docids is undefined

yvaucher commented on l10n_ch_payment_slip/report/ir_action.py:
```diff
@@ -1,46 +1,201 @@
+        if (self.name == 'l10n_ch_payment_slip.'
+                         'one_slip_per_page_from_invoice'):
+            report = self._get_report_from_name(self.name)
+            save_in_attachment = self._check_attachment_use(
```
_check_attachment_use does not exist anymore in Odoo 11.0.

This was removed here 

https://github.com/odoo/odoo/commit/e80238042c9d93d492ea8b06b0041aced0d81dcd#diff-c941fd9fa97231fc06b29633532dd307L318

yvaucher committed:
Migration of attachment feature to v11

grindtildeath committed:
Fix dependency to remove std print isr button

Commit:
Fix right error on ISR print

yvaucher committed:
Merge pull request #5 from p-tombez/11.0-mig-l10n_ch_payment_slip

Fix right error on ISR print

BT-ojossen commented:
hi, what's the state here? when will this be merged?

yvaucher commented:
@BT-ojossen this is in a ready state, but for the unit tests which are still red. Thus to answer your question, it will be merged when the tests are fixed.

grindtildeath committed:
Finish migration to 11.0

Restore report_type to reportlab-pdf

Improvement to batch wizard, error readonly

grindtildeath committed:
Fix merges methods

BT-ccebrian commented on l10n_ch_payment_slip/wizard/isr_batch_print.py:
```diff
@@ -1,22 +1,21 @@
 
     invoice_ids = fields.Many2many(comodel_name='account.invoice',
                                    string='Invoices')
     error_message = fields.Text('Errors')
```
Hi @yvaucher, I think it is a silly mistake but,

When you get the related error in the batch print wizard, the error is editable, in my point of view it is a little bit annoying.

Regards!

yvaucher replied:
@BT-ccebrian makes sense to me.

@grindtildeath can you also add a readonly on that field? AFAIK it should only be set server side.

grindtildeath replied:
Fixed in dee27ae205486dfc0b8e6f217a657567086ed1e3

grindtildeath committed:
Rework save as attachment feature using _post_pdf

grindtildeath committed:
Drop unneeded imports

grindtildeath committed:
Finally fix the test considering odoo weird behaviour

simahawk reviewed (approved):
LG.
Runbot is not really red, is yellow because of warnings unrelated to this module.

One thing: we might want to mention in a TODO that the merge mode machinery could/should be splitted into reporting-engine module(s)

yvaucher commented:
@grindtildeath LG thanks for fixing the tests

grindtildeath reviewed (approved):
I approve as I'm the author of the last commits :blush:
